clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,411 @@
+		pc.stop()
+	}
+	s.pingCheckers = make(map[string]*pingChecker)
+}
+
+// String returns a string representation of the metric.
+func (s *serviceCreationLatencyMeasurement) String() string {
+	return serviceCreationLatencyName + ": " + s.selectorsString
+}
+
+type serviceCreationTimes struct {
+	created    time.Time
+	ipAssigned time.Time
+	reachable  time.Time
+}
+
+func (s *serviceCreationLatencyMeasurement) start() error {
+	s.selectorsString = measurementutil.CreateSelectorsString(s.namespace, s.labelSelector, s.fieldSelector)
+	if s.isRunning {
+		klog.Infof("%s: wait for controlled pods measurement already running", s)
+		return nil
+	}
+	klog.Infof("%s: starting wait for controlled pods measurement...", s)
+
+	s.isRunning = true
+	s.stopCh = make(chan struct{})
+
+	i := informer.NewInformer(
+		s.client,
+		"services",
+		s.namespace,
+		s.fieldSelector,
+		s.labelSelector,
+		func(oldObj, newObj interface{}) {
+			f := func() {
+				s.handleObject(oldObj, newObj)
+			}
+			s.queue.Add(&f)
+		},
+	)
+	return informer.StartAndSync(i, s.stopCh, informerSyncTimeout)
+}
+
+func (s *serviceCreationLatencyMeasurement) waitForReady() error {
+	for {
+		created, ipAssigned, reachable := 0, 0, 0
+		select {
+		case <-time.After(defaultWaitTimeout):
+			klog.Errorf("%s: waiting for services timeout")
+			return fmt.Errorf("waiting for services timeout")
+		case <-time.After(defaultCheckInterval):
+			func() {
+				s.lock.Lock()
+				defer s.lock.Unlock()
+				for _, ct := range s.creationTimes {
+					created++
+					if !ct.ipAssigned.IsZero() {
+						ipAssigned++
+					}
+					if !ct.reachable.IsZero() {
+						reachable++
+					}
+				}
+			}()
+			klog.Infof("%s: %d created, %d ipAssigned, %d reachable", s, created, ipAssigned, reachable)
+			if created == reachable {
+				return nil
+			}
+		}
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering service created latency measurement...", s)
+	if !s.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", s)
+	}
+
+	createToAssignedLatency := make([]measurementutil.LatencyData, 0)
+	assignedToAvailableLatency := make([]measurementutil.LatencyData, 0)
+	createToAvailableLatency := make([]measurementutil.LatencyData, 0)
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for key, ct := range s.creationTimes {
+		if ct.created.IsZero() || ct.ipAssigned.IsZero() {
+			continue
+		}
+		createToAssignedLatency = append(createToAssignedLatency, serviceCreationLatencyData{
+			Key:     key,
+			Latency: ct.ipAssigned.Sub(ct.created),
+		})
+		if ct.reachable.IsZero() {
+			continue
+		}
+		assignedToAvailableLatency = append(assignedToAvailableLatency, serviceCreationLatencyData{
+			Key:     key,
+			Latency: ct.reachable.Sub(ct.ipAssigned),
+		})
+		createToAvailableLatency = append(createToAvailableLatency, serviceCreationLatencyData{
+			Key:     key,
+			Latency: ct.reachable.Sub(ct.created),
+		})
+	}
```

Comment:
This whole stuff is boilerplate - I've been already thinking how to make it cleaner and avoid doing the same stuff in every single measurement, and I think that I have some idea now.

So here is the idea:
"creationTimes" (and similar things in other measurements like podStartupTime) should be of the type of:
map[string]time.Time [means phase->timestamp]
[Then this type is common for every single measurement, only names of phases differ]

- in the individual measurement you only define names of phases and how to compute timestamps of them for individual object

- then you have a generic "Compute" function (well, we need better name) as following:
func (*objectTransitionTimes) Compute(map[string]pair<string,string>) map[string]measurementutil.LatencyMetric {}

that basically computes for every given string->pair<phase1,phase2>
- first list of latencies of all objects between phase1->phase2
- sorts them
- prints latencies
- and computes percentiles

Then you basically write this logic once and will be able to reuse it in every measurement producing latency.
[This will also introduce a common pattern for writing such measurements, which I'm really looking for]

Reply:
You can first do it for pod_startup_latency to get framework (in a separate PR) and then use it here.

Reply:
Done.